samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java

```diff
@@ -280,4 +276,50 @@ public void onBarrierError(String version, Throwable t) {
       stop();
     }
   }
+
+  /// listener to handle session expiration
+  public class ZkSessionStateChangedListener implements IZkStateListener {
+
+    @Override
+    public void handleStateChanged(Watcher.Event.KeeperState state)
+        throws Exception {
+      if (state == Watcher.Event.KeeperState.Expired) {
+        // if the session has expired it means that all the registration's ephemeral nodes are gone.
+        LOG.warn("Got session expired event for processor=" + processorId);
+      }
+    }
+
+    @Override
+    public void handleNewSession()
+        throws Exception {
+      LOG.info("Got new session created event for processor=" + processorId);
+
+      // increase generation of the ZK connection. All the callbacks from the previous generation will be ignored.
+      zkUtils.incGeneration();
+
+      if (coordinatorListener != null) {
+        coordinatorListener.onJobModelExpired();
+      }
+      // reset all the values that might've been from the previous session (e.g ephemeral node path)
+      zkUtils.unregister();
+
+      if (zkController != null) {
+        LOG.info("register zk controller for the new session");
+        zkController.register();
+      }
+    }
+
+    @Override
+    public void handleSessionEstablishmentError(Throwable error)
+        throws Exception {
+      // this means we cannot connect to zookeeper
+      LOG.info("=================================================handleSessionEstablishmentError received for processor=" + processorId, error);
+      stop();
```

Comment:
No, we don't stop debounce on a new Session establishment. We stop() everything if we get sessionEstablishmentError - which means client couldn't re-establish a session and it is a hard failure.